[Flaking tests]Test_allocateWithNominated/reservation-ignored_pod panic with nil pointer dereference

### Description
﻿
`Test_allocateWithNominated/reservation-ignored_pod` in `pkg/scheduler/plugins/deviceshare/reservation_test.go` panics with a nil pointer dereference at `devicehandler_gpu.go:41`.
﻿
This test case was introduced in commit `72af3153` (PR #2752: scheduler: deviceshare adapt pre-allocation) on 2026-01-06.
﻿
### Steps to Reproduce
﻿
go test ./pkg/scheduler/plugins/deviceshare/... -run "Test_allocateWithNominated/reservation-ignored_pod" -v
﻿
Panic Stack Trace
﻿
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x30f54ed]
﻿
Failed Ci :https://github.com/koordinator-sh/koordinator/actions/runs/22379282089/job/64776244495?pr=2799
﻿
Root Cause
﻿
In the reservation-ignored pod test case, allocateWithNominated calls tryAllocateIgnoreReservation, which calls allocator.Allocate() → Prepare() → calcRequestsAndCountByDeviceType() → GPUHandler.CalcDesiredRequestsAndCount(). At devicehandler_gpu.go:41, it accesses nodeDevice.deviceTotal[schedulingv1alpha1.GPU] but nodeDevice is nil, causing the panic.

--- a/pkg/scheduler/apis/config/types.go
+++ b/pkg/scheduler/apis/config/types.go
@@ -210,6 +210,9 @@ type ReservationArgs struct {
 	// DisableGarbageCollection indicates whether to disable garbage collection.
 	// Default to false if unspecified.
 	DisableGarbageCollection bool
+	// ResyncIntervalSeconds is the duration in seconds between each turns of resync.
+	// Defaults to 60 seconds if unspecified.
+	ResyncIntervalSeconds int64
 }
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

--- a/pkg/scheduler/apis/config/v1/defaults.go
+++ b/pkg/scheduler/apis/config/v1/defaults.go
@@ -58,6 +58,7 @@ var (
 	defaultReservationControllerWorkers = ptr.To[int32](1)
 	defaultGCDurationSeconds            = ptr.To[int64](86400)
 	defaultGCIntervalSeconds            = ptr.To[int64](60)
+	defaultResyncIntervalSeconds        = ptr.To[int64](60)
 
 	defaultDelayEvictTime       = 120 * time.Second
 	defaultRevokePodInterval    = 1 * time.Second
@@ -179,6 +180,9 @@ func SetDefaults_ReservationArgs(obj *ReservationArgs) {
 	}
 	if obj.GCIntervalSeconds == 0 {
 		obj.GCIntervalSeconds = *defaultGCIntervalSeconds
+	}
+	if obj.ResyncIntervalSeconds == 0 {
+		obj.ResyncIntervalSeconds = *defaultResyncIntervalSeconds
 	}
 }
 

--- a/pkg/scheduler/apis/config/v1/types.go
+++ b/pkg/scheduler/apis/config/v1/types.go
@@ -205,6 +205,9 @@ type ReservationArgs struct {
 	// DisableGarbageCollection indicates whether to disable garbage collection.
 	// Default to false if unspecified.
 	DisableGarbageCollection bool `json:"disableGarbageCollection,omitempty"`
+	// ResyncIntervalSeconds is the duration in seconds between each turns of resync.
+	// Defaults to 60 seconds if unspecified.
+	ResyncIntervalSeconds int64 `json:"resyncIntervalSeconds,omitempty"`
 }
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

--- a/pkg/scheduler/apis/config/v1/zz_generated.conversion.go
+++ b/pkg/scheduler/apis/config/v1/zz_generated.conversion.go
@@ -554,6 +554,7 @@ func autoConvert_v1_ReservationArgs_To_config_ReservationArgs(in *ReservationArg
 	out.GCDurationSeconds = in.GCDurationSeconds
 	out.GCIntervalSeconds = in.GCIntervalSeconds
 	out.DisableGarbageCollection = in.DisableGarbageCollection
+	out.ResyncIntervalSeconds = in.ResyncIntervalSeconds
 	return nil
 }
 
@@ -578,6 +579,7 @@ func autoConvert_config_ReservationArgs_To_v1_ReservationArgs(in *config.Reserva
 	out.GCDurationSeconds = in.GCDurationSeconds
 	out.GCIntervalSeconds = in.GCIntervalSeconds
 	out.DisableGarbageCollection = in.DisableGarbageCollection
+	out.ResyncIntervalSeconds = in.ResyncIntervalSeconds
 	return nil
 }
 

--- a/pkg/scheduler/apis/config/v1beta3/defaults.go
+++ b/pkg/scheduler/apis/config/v1beta3/defaults.go
@@ -58,6 +58,7 @@ var (
 	defaultReservationControllerWorkers = ptr.To[int32](1)
 	defaultGCDurationSeconds            = ptr.To[int64](86400)
 	defaultGCIntervalSeconds            = ptr.To[int64](60)
+	defaultResyncIntervalSeconds        = ptr.To[int64](60)
 
 	defaultDelayEvictTime       = 120 * time.Second
 	defaultRevokePodInterval    = 1 * time.Second
@@ -179,6 +180,9 @@ func SetDefaults_ReservationArgs(obj *ReservationArgs) {
 	}
 	if obj.GCIntervalSeconds == 0 {
 		obj.GCIntervalSeconds = *defaultGCIntervalSeconds
+	}
+	if obj.ResyncIntervalSeconds == 0 {
+		obj.ResyncIntervalSeconds = *defaultResyncIntervalSeconds
 	}
 }
 

--- a/pkg/scheduler/apis/config/v1beta3/types.go
+++ b/pkg/scheduler/apis/config/v1beta3/types.go
@@ -206,6 +206,9 @@ type ReservationArgs struct {
 	// DisableGarbageCollection indicates whether to disable garbage collection.
 	// Default to false if unspecified.
 	DisableGarbageCollection bool `json:"disableGarbageCollection,omitempty"`
+	// ResyncIntervalSeconds is the duration in seconds between each turns of resync.
+	// Defaults to 60 seconds if unspecified.
+	ResyncIntervalSeconds int64 `json:"resyncIntervalSeconds,omitempty"`
 }
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

--- a/pkg/scheduler/apis/config/v1beta3/zz_generated.conversion.go
+++ b/pkg/scheduler/apis/config/v1beta3/zz_generated.conversion.go
@@ -549,6 +549,7 @@ func autoConvert_v1beta3_ReservationArgs_To_config_ReservationArgs(in *Reservati
 	out.GCDurationSeconds = in.GCDurationSeconds
 	out.GCIntervalSeconds = in.GCIntervalSeconds
 	out.DisableGarbageCollection = in.DisableGarbageCollection
+	out.ResyncIntervalSeconds = in.ResyncIntervalSeconds
 	return nil
 }
 
@@ -573,6 +574,7 @@ func autoConvert_config_ReservationArgs_To_v1beta3_ReservationArgs(in *config.Re
 	out.GCDurationSeconds = in.GCDurationSeconds
 	out.GCIntervalSeconds = in.GCIntervalSeconds
 	out.DisableGarbageCollection = in.DisableGarbageCollection
+	out.ResyncIntervalSeconds = in.ResyncIntervalSeconds
 	return nil
 }
 

--- a/pkg/scheduler/apis/config/validation/validation_pluginargs.go
+++ b/pkg/scheduler/apis/config/validation/validation_pluginargs.go
@@ -253,6 +253,14 @@ func ValidateReservationArgs(path *field.Path, args *config.ReservationArgs) err
 		))
 	}
 
+	if args.ResyncIntervalSeconds < 0 {
+		allErrs = append(allErrs, field.Invalid(
+			path.Child("ResyncIntervalSeconds"),
+			args.ResyncIntervalSeconds,
+			"must be non-negative",
+		))
+	}
+
 	if len(allErrs) == 0 {
 		return nil
 	}

--- a/pkg/scheduler/metrics/metrics.go
+++ b/pkg/scheduler/metrics/metrics.go
@@ -21,7 +21,6 @@ import (
 	"time"
 
 	"github.com/prometheus/client_golang/prometheus"
-
 	"k8s.io/component-base/metrics"
 	"k8s.io/component-base/metrics/legacyregistry"
 	schedulermetrics "k8s.io/kubernetes/pkg/scheduler/metrics"
@@ -42,12 +41,12 @@ var (
 			Help:           "The currently scheduled Pod exceeds the maximum acceptable time interval",
 			StabilityLevel: metrics.STABLE,
 		}, []string{"profile"})
-	ReservationStatusPhase = utilmetrics.NewGCGaugeVec("reservation_status_phase", prometheus.NewGaugeVec(
-		prometheus.GaugeOpts{
+	ReservationStatusPhase = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
 			Subsystem: schedulermetrics.SchedulerSubsystem,
 			Name:      "reservation_status_phase",
 			Help:      `The current number of reservations in each status phase (e.g. Pending, Available, Succeeded, Failed)`,
-		}, []string{"name", "phase"}))
+		}, []string{"name", "phase"})
 	ReservationResource = utilmetrics.NewGCGaugeVec(
 		"reservation_resource",
 		prometheus.NewGaugeVec(
@@ -127,6 +126,7 @@ var (
 
 	metricsList = []metrics.Registerable{
 		SchedulingTimeout,
+		ReservationStatusPhase,
 		PodSchedulingEvaluatedNodes,
 		PodSchedulingFeasibleNodes,
 		ElasticQuotaProcessLatency,
@@ -137,7 +137,6 @@ var (
 	}
 
 	gcMetricsList = []prometheus.Collector{
-		ReservationStatusPhase.GetGaugeVec(),
 		ReservationResource.GetGaugeVec(),
 		JobPreemptionDuration.GetHistogramVec(),
 	}
@@ -164,27 +163,6 @@ const (
 	TypeUtilization = "utilization"
 )
 
-// RecordReservationPhase records the phase of a reservation as a metric.
-// It uses the provided name, phase, and value to set the metric with specific labels.
-func RecordReservationPhase(name string, phase string, value float64) {
-	labels := prometheus.Labels{
-		reservationNameKey:  name,
-		reservationPhaseKey: phase,
-	}
-	ReservationStatusPhase.WithSet(labels, value)
-}
-
-// RecordReservationResourceByTypeWithUnit records the resource record of a reservation as a metric.
-func RecordReservationResourceByTypeWithUnit(name, resource, typ, unit string, value float64) {
-	labels := prometheus.Labels{
-		reservationResourceTypeKey: typ,
-		reservationNameKey:         name,
-		reservationResourceKey:     resource,
-		reservationResourceUnitKey: unit,
-	}
-	ReservationResource.WithSet(labels, value)
-}
-
 var registerMetrics sync.Once
 
 // Register all metrics.
@@ -209,6 +187,31 @@ func RegisterGCMetrics(gcMetrics ...prometheus.Collector) {
 	for _, metric := range gcMetrics {
 		legacyregistry.RawMustRegister(metric)
 	}
+}
+
+// RecordReservationPhase records the phase of a reservation as a metric.
+// It uses the provided name, phase, and value to set the metric with specific labels.
+func RecordReservationPhase(name string, phase string, value float64) {
+	labels := prometheus.Labels{
+		reservationNameKey:  name,
+		reservationPhaseKey: phase,
+	}
+	ReservationStatusPhase.With(labels).Set(value)
+}
+
+func ResetReservationPhase() {
+	ReservationStatusPhase.Reset()
+}
+
+// RecordReservationResourceByTypeWithUnit records the resource record of a reservation as a metric.
+func RecordReservationResourceByTypeWithUnit(name, resource, typ, unit string, value float64) {
+	labels := prometheus.Labels{
+		reservationResourceTypeKey: typ,
+		reservationNameKey:         name,
+		reservationResourceKey:     resource,
+		reservationResourceUnitKey: unit,
+	}
+	ReservationResource.WithSet(labels, value)
 }
 
 func RecordElasticQuotaProcessLatency(operation string, latency time.Duration) {

--- a/pkg/scheduler/plugins/deviceshare/reservation_test.go
+++ b/pkg/scheduler/plugins/deviceshare/reservation_test.go
@@ -1107,53 +1107,60 @@ func Test_tryAllocateFromReservation(t *testing.T) {
 }
 
 func Test_allocateWithNominated(t *testing.T) {
-	suit := newPluginTestSuit(t, nil)
-	p, err := suit.proxyNew(getDefaultArgs(), suit.Framework)
-	assert.NoError(t, err)
-	pl := p.(*Plugin)
+	// Fixed UIDs to avoid non-determinism from uuid.NewUUID()
+	const (
+		reservationUID = types.UID("test-reservation-uid")
+		reservePodUID  = types.UID("test-reserve-pod-uid")
+		ignoredPodUID  = types.UID("test-ignored-pod-uid")
+		normalPodUID   = types.UID("test-normal-pod-uid")
+	)
 
-	// Setup node device
-	device := &schedulingv1alpha1.Device{
-		Spec: schedulingv1alpha1.DeviceSpec{
-			Devices: []schedulingv1alpha1.DeviceInfo{
-				{
-					Type:   schedulingv1alpha1.GPU,
-					Minor:  ptr.To[int32](0),
-					Health: true,
-					Resources: corev1.ResourceList{
-						apiext.ResourceGPUCore:        resource.MustParse("100"),
-						apiext.ResourceGPUMemory:      resource.MustParse("8Gi"),
-						apiext.ResourceGPUMemoryRatio: resource.MustParse("100"),
+	// newDevice returns a fresh Device with two healthy GPUs (minor 0 and 1).
+	newDevice := func() *schedulingv1alpha1.Device {
+		return &schedulingv1alpha1.Device{
+			Spec: schedulingv1alpha1.DeviceSpec{
+				Devices: []schedulingv1alpha1.DeviceInfo{
+					{
+						Type:   schedulingv1alpha1.GPU,
+						Minor:  ptr.To[int32](0),
+						Health: true,
+						Resources: corev1.ResourceList{
+							apiext.ResourceGPUCore:        resource.MustParse("100"),
+							apiext.ResourceGPUMemory:      resource.MustParse("8Gi"),
+							apiext.ResourceGPUMemoryRatio: resource.MustParse("100"),
+						},
 					},
-				},
-				{
-					Type:   schedulingv1alpha1.GPU,
-					Minor:  ptr.To[int32](1),
-					Health: true,
-					Resources: corev1.ResourceList{
-						apiext.ResourceGPUCore:        resource.MustParse("100"),
-						apiext.ResourceGPUMemory:      resource.MustParse("8Gi"),
-						apiext.ResourceGPUMemoryRatio: resource.MustParse("100"),
+					{
+						Type:   schedulingv1alpha1.GPU,
+						Minor:  ptr.To[int32](1),
+						Health: true,
+						Resources: corev1.ResourceList{
+							apiext.ResourceGPUCore:        resource.MustParse("100"),
+							apiext.ResourceGPUMemory:      resource.MustParse("8Gi"),
+							apiext.ResourceGPUMemoryRatio: resource.MustParse("100"),
+						},
 					},
 				},
 			},
-		},
+		}
 	}
-	pl.nodeDeviceCache.updateNodeDevice("test-node", device)
 
-	reservation := &schedulingv1alpha1.Reservation{
-		ObjectMeta: metav1.ObjectMeta{
-			UID:  uuid.NewUUID(),
-			Name: "test-reservation",
-		},
-		Spec: schedulingv1alpha1.ReservationSpec{
-			Template: &corev1.PodTemplateSpec{},
-		},
-		Status: schedulingv1alpha1.ReservationStatus{
-			NodeName: "test-node",
-		},
+	// newReservationInfo returns a fresh ReservationInfo with a fixed UID.
+	newReservationInfo := func() *frameworkext.ReservationInfo {
+		reservation := &schedulingv1alpha1.Reservation{
+			ObjectMeta: metav1.ObjectMeta{
+				UID:  reservationUID,
+				Name: "test-reservation",
+			},
+			Spec: schedulingv1alpha1.ReservationSpec{
+				Template: &corev1.PodTemplateSpec{},
+			},
+			Status: schedulingv1alpha1.ReservationStatus{
+				NodeName: "test-node",
+			},
+		}
+		return frameworkext.NewReservationInfo(reservation)
 	}
-	rInfo := frameworkext.NewReservationInfo(reservation)
 
 	node := &corev1.Node{
 		ObjectMeta: metav1.ObjectMeta{
@@ -1162,62 +1169,66 @@ func Test_allocateWithNominated(t *testing.T) {
 	}
 
 	tests := []struct {
-		name         string
-		pod          *corev1.Pod
-		restoreState *nodeReservationRestoreStateData
-		wantNil      bool
-		wantSuccess  bool
+		name              string
+		pod               *corev1.Pod
+		buildRestoreState func(rInfo *frameworkext.ReservationInfo) *nodeReservationRestoreStateData
+		wantNil           bool
+		wantSuccess       bool
 	}{
 		{
 			name: "reserve pod without pre-allocation",
 			pod: &corev1.Pod{
 				ObjectMeta: metav1.ObjectMeta{
 					Name: "reserve-pod",
-					UID:  uuid.NewUUID(),
+					UID:  reservePodUID,
 					Annotations: map[string]string{
 						reservationutil.AnnotationReservePod: "true",
 					},
 				},
 			},
-			restoreState: &nodeReservationRestoreStateData{},
-			wantNil:      true,
-			wantSuccess:  true,
+			buildRestoreState: func(_ *frameworkext.ReservationInfo) *nodeReservationRestoreStateData {
+				return &nodeReservationRestoreStateData{}
+			},
+			wantNil:     true,
+			wantSuccess: true,
 		},
 		{
 			name: "reservation-ignored pod",
 			pod: &corev1.Pod{
 				ObjectMeta: metav1.ObjectMeta{
 					Name: "ignored-pod",
-					UID:  uuid.NewUUID(),
+					UID:  ignoredPodUID,
 					Labels: map[string]string{
 						apiext.LabelReservationIgnored: "true",
 					},
 				},
 			},
-			restoreState: &nodeReservationRestoreStateData{
-				matched: []reusableAlloc{
-					{
-						rInfo: rInfo,
-						allocatable: map[schedulingv1alpha1.DeviceType]deviceResources{
-							schedulingv1alpha1.GPU: {
-								0: {
-									apiext.ResourceGPUCore:        resource.MustParse("50"),
-									apiext.ResourceGPUMemory:      resource.MustParse("4Gi"),
-									apiext.ResourceGPUMemoryRatio: resource.MustParse("50"),
+			buildRestoreState: func(rInfo *frameworkext.ReservationInfo) *nodeReservationRestoreStateData {
+				return &nodeReservationRestoreStateData{
+					matched: []reusableAlloc{
+						{
+							rInfo: rInfo,
+							allocatable: map[schedulingv1alpha1.DeviceType]deviceResources{
+								schedulingv1alpha1.GPU: {
+									0: {
+										apiext.ResourceGPUCore:        resource.MustParse("50"),
+										apiext.ResourceGPUMemory:      resource.MustParse("4Gi"),
+										apiext.ResourceGPUMemoryRatio: resource.MustParse("50"),
+									},
 								},
 							},
-						},
-						remained: map[schedulingv1alpha1.DeviceType]deviceResources{
-							schedulingv1alpha1.GPU: {
-								0: {
-									apiext.ResourceGPUCore:        resource.MustParse("50"),
-									apiext.ResourceGPUMemory:      resource.MustParse("4Gi"),
-									apiext.ResourceGPUMemoryRatio: resource.MustParse("50"),
+							remained: map[schedulingv1alpha1.DeviceType]deviceResources{
+								schedulingv1alpha1.GPU: {
+									0: {
+										apiext.ResourceGPUCore:        resource.MustParse("50"),
+										apiext.ResourceGPUMemory:      resource.MustParse("4Gi"),
+										apiext.ResourceGPUMemoryRatio: resource.MustParse("50"),
+									},
 								},
 							},
 						},
 					},
-				},
+				}
 			},
 			wantNil:     false,
 			wantSuccess: true,
@@ -1227,17 +1238,32 @@ func Test_allocateWithNominated(t *testing.T) {
 			pod: &corev1.Pod{
 				ObjectMeta: metav1.ObjectMeta{
 					Name: "normal-pod",
-					UID:  uuid.NewUUID(),
+					UID:  normalPodUID,
 				},
 			},
-			restoreState: &nodeReservationRestoreStateData{},
-			wantNil:      true,
-			wantSuccess:  true,
+			buildRestoreState: func(_ *frameworkext.ReservationInfo) *nodeReservationRestoreStateData {
+				return &nodeReservationRestoreStateData{}
+			},
+			wantNil:     true,
+			wantSuccess: true,
 		},
 	}
 
 	for _, tt := range tests {
+		tt := tt
 		t.Run(tt.name, func(t *testing.T) {
+			// Each subtest gets its own isolated plugin instance, device cache, and rInfo
+			// to avoid any shared-state races between subtests or background goroutines.
+			suit := newPluginTestSuit(t, nil)
+			p, err := suit.proxyNew(getDefaultArgs(), suit.Framework)
+			assert.NoError(t, err)
+			pl := p.(*Plugin)
+
+			pl.nodeDeviceCache.updateNodeDevice("test-node", newDevice())
+
+			rInfo := newReservationInfo()
+			restoreState := tt.buildRestoreState(rInfo)
+
 			state := &preFilterState{
 				podRequests: map[schedulingv1alpha1.DeviceType]corev1.ResourceList{
 					schedulingv1alpha1.GPU: {
@@ -1245,9 +1271,10 @@ func Test_allocateWithNominated(t *testing.T) {
 						apiext.ResourceGPUMemory: resource.MustParse("4Gi"),
 					},
 				},
-				preemptibleInRRs: map[string]map[types.UID]map[schedulingv1alpha1.DeviceType]deviceResources{},
+				preemptibleInRRs: map[string]map[types.UID]map[schedulingv1alpha1.DeviceType]deviceResources{
+					"test-node": {},
+				},
 			}
-			state.preemptibleInRRs["test-node"] = map[types.UID]map[schedulingv1alpha1.DeviceType]deviceResources{}
 			state.gpuRequirements, _ = parseGPURequirements(tt.pod, state.podRequests, nil, nil, nil)
 
 			nodeDeviceInfo := pl.nodeDeviceCache.getNodeDevice("test-node", false)
@@ -1261,7 +1288,7 @@ func Test_allocateWithNominated(t *testing.T) {
 			result, status := pl.allocateWithNominated(
 				allocator,
 				state,
-				tt.restoreState,
+				restoreState,
 				node,
 				tt.pod,
 				nil,

--- a/pkg/scheduler/plugins/reservation/controller/controller.go
+++ b/pkg/scheduler/plugins/reservation/controller/controller.go
@@ -27,6 +27,7 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/types"
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	"k8s.io/apimachinery/pkg/util/wait"
@@ -76,6 +77,7 @@ type Controller struct {
 	isGCDisabled               bool
 	gcDuration                 time.Duration
 	gcInterval                 time.Duration
+	resyncInterval             time.Duration
 
 	lock   sync.RWMutex
 	pods   map[string]map[types.UID]*corev1.Pod    // nodeName -> podUID -> pod
@@ -114,6 +116,10 @@ func New(
 	if args != nil && args.GCIntervalSeconds > 0 {
 		gcInterval = time.Duration(args.GCIntervalSeconds) * time.Second
 	}
+	resyncInterval := defaultResyncInterval
+	if args != nil && args.ResyncIntervalSeconds > 0 {
+		resyncInterval = time.Duration(args.ResyncIntervalSeconds) * time.Second
+	}
 	return &Controller{
 		sharedInformerFactory:      sharedInformerFactory,
 		koordSharedInformerFactory: koordSharedInformerFactory,
@@ -127,6 +133,7 @@ func New(
 		isGCDisabled:               isGCDisabled,
 		gcDuration:                 gcDuration,
 		gcInterval:                 gcInterval,
+		resyncInterval:             resyncInterval,
 		pods:                       map[string]map[types.UID]*corev1.Pod{},
 		podToR:                     map[types.UID]types.UID{},
 		rToPod:                     map[types.UID]map[types.UID]*corev1.Pod{},
@@ -168,6 +175,11 @@ func (c *Controller) Start() {
 		go wait.Until(c.gcReservations, c.gcInterval, nil)
 	} else {
 		klog.V(4).InfoS("garbage collection for reservations is disabled")
+	}
+	if c.resyncInterval > 0 {
+		go wait.Until(c.resyncReservations, c.resyncInterval, nil)
+	} else {
+		klog.V(4).InfoS("resync for reservations is disabled")
 	}
 }
 
@@ -320,7 +332,6 @@ func (c *Controller) syncStatus(reservation *schedulingv1alpha1.Reservation, pod
 	}
 
 	if reservation.Status.NodeName == "" {
-		RecordReservationPhases(reservation)
 		return nil
 	}
 
@@ -365,7 +376,6 @@ func (c *Controller) syncStatus(reservation *schedulingv1alpha1.Reservation, pod
 }
 
 func (c *Controller) updateReservationStatus(reservation *schedulingv1alpha1.Reservation) error {
-	RecordReservationPhases(reservation)
 	_, err := c.koordClientSet.SchedulingV1alpha1().Reservations().UpdateStatus(context.TODO(), reservation, metav1.UpdateOptions{})
 	if err != nil {
 		klog.ErrorS(err, "Failed to update reservation status", "reservation", klog.KObj(reservation), "uid", reservation.UID, "phase", reservation.Status.Phase)
@@ -408,6 +418,24 @@ func nextSyncTime(r *schedulingv1alpha1.Reservation) time.Duration {
 		duration = maxRetryAfterTime
 	}
 	return duration
+}
+
+const (
+	defaultResyncInterval = 60 * time.Second
+)
+
+func (c *Controller) resyncReservations() {
+	reservations, err := c.reservationLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("failed to list reservations, abort the resync turn, err: %s", err)
+		return
+	}
+	metrics.ResetReservationPhase()
+	for _, reservation := range reservations {
+		// record metrics
+		RecordReservationPhases(reservation)
+	}
+	klog.V(4).InfoS("resynced reservation metrics", "count", len(reservations))
 }
 
 // RecordReservationPhases records all possible phases of a reservation as metrics.

--- a/pkg/scheduler/plugins/reservation/controller/controller_test.go
+++ b/pkg/scheduler/plugins/reservation/controller/controller_test.go
@@ -29,12 +29,15 @@ import (
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/util/wait"
 	k8sfeature "k8s.io/apiserver/pkg/util/feature"
 	"k8s.io/client-go/informers"
 	kubefake "k8s.io/client-go/kubernetes/fake"
 	basemetrics "k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
 	"k8s.io/component-base/metrics/testutil"
 	"k8s.io/utils/ptr"
 
@@ -226,12 +229,11 @@ func TestSyncStatus(t *testing.T) {
 	sharedInformerFactory := informers.NewSharedInformerFactory(fakeClientSet, 0)
 	koordSharedInformerFactory := koordinformers.NewSharedInformerFactory(fakeKoordClientSet, 0)
 	// register metrics
-	metricsRegistry := basemetrics.NewKubeRegistry()
-	metrics.ReservationStatusPhase.GetGaugeVec().Reset()
+	metrics.Register()
+	metrics.ReservationStatusPhase.Reset()
+	metricsResourceRegistry := basemetrics.NewKubeRegistry()
 	metrics.ReservationResource.GetGaugeVec().Reset()
-	metricsRegistry.Registerer().MustRegister(
-		metrics.ReservationStatusPhase.GetGaugeVec(),
-		metrics.ReservationResource.GetGaugeVec())
+	metricsResourceRegistry.Registerer().MustRegister(metrics.ReservationResource.GetGaugeVec())
 
 	reservation := &schedulingv1alpha1.Reservation{
 		ObjectMeta: metav1.ObjectMeta{
@@ -334,27 +336,8 @@ func TestSyncStatus(t *testing.T) {
 		cond.LastTransitionTime = metav1.Time{}
 	}
 	assert.Equal(t, expectReservation, got)
-	// check metrics
-	expectAvailableMetricCount := 0
-	expectFailedMetricCount := 0
-	expectPendingMetricCount := 0
-	expectSucceededMetricCount := 1
-	expectedSuccessedMetrics := fmt.Sprintf(`
-				# HELP scheduler_reservation_status_phase The current number of reservations in each status phase (e.g. Pending, Available, Succeeded, Failed)
-				# TYPE scheduler_reservation_status_phase gauge
-				scheduler_reservation_status_phase{name="%s",phase="Available"} %v
-				scheduler_reservation_status_phase{name="%s",phase="Failed"} %v
-				scheduler_reservation_status_phase{name="%s",phase="Pending"} %v
-				scheduler_reservation_status_phase{name="%s",phase="Succeeded"} %v
-				`,
-		"normalReservation", expectAvailableMetricCount,
-		"normalReservation", expectFailedMetricCount,
-		"normalReservation", expectPendingMetricCount,
-		"normalReservation", expectSucceededMetricCount,
-	)
-	if err := testutil.GatherAndCompare(metricsRegistry, strings.NewReader(expectedSuccessedMetrics), "scheduler_reservation_status_phase"); err != nil {
-		t.Error(err)
-	}
+	// phase metrics are only written by resyncReservations(), not by syncStatus();
+	// no phase metric assertion here because the legacyregistry is shared across tests.
 
 	expectedUtilizationMetrics := fmt.Sprintf(`
 # HELP scheduler_reservation_resource Resource metrics for a reservation, including allocatable, allocated, and utilization with unit.
@@ -371,7 +354,7 @@ scheduler_reservation_resource{name="%s",resource="memory",type="utilization",un
 		"normalReservation", "normalReservation",
 	)
 
-	if err := testutil.GatherAndCompare(metricsRegistry, strings.NewReader(expectedUtilizationMetrics),
+	if err := testutil.GatherAndCompare(metricsResourceRegistry, strings.NewReader(expectedUtilizationMetrics),
 		"scheduler_reservation_resource"); err != nil {
 		t.Error(err)
 	}
@@ -780,6 +763,202 @@ func TestPodEventHandlerUnassignedPod(t *testing.T) {
 	controller.onPodUpdate(oldPod, newPod)
 	assert.Equal(t, 0, len(controller.getPodsOnReservation("test-reservation-uid")))
 	assert.Equal(t, 0, len(controller.getPodsOnNode("test-node-1")))
+}
+
+func TestRecordReservationPhases(t *testing.T) {
+	tests := []struct {
+		name            string
+		reservation     *schedulingv1alpha1.Reservation
+		expectedMetrics string
+	}{
+		{
+			name: "Available reservation",
+			reservation: &schedulingv1alpha1.Reservation{
+				ObjectMeta: metav1.ObjectMeta{Name: "r-available"},
+				Status: schedulingv1alpha1.ReservationStatus{
+					Phase: schedulingv1alpha1.ReservationAvailable,
+				},
+			},
+			expectedMetrics: `
+# HELP scheduler_reservation_status_phase [ALPHA] The current number of reservations in each status phase (e.g. Pending, Available, Succeeded, Failed)
+# TYPE scheduler_reservation_status_phase gauge
+scheduler_reservation_status_phase{name="r-available",phase="Available"} 1
+scheduler_reservation_status_phase{name="r-available",phase="Failed"} 0
+scheduler_reservation_status_phase{name="r-available",phase="Pending"} 0
+scheduler_reservation_status_phase{name="r-available",phase="Succeeded"} 0
+`,
+		},
+		{
+			name: "Pending reservation",
+			reservation: &schedulingv1alpha1.Reservation{
+				ObjectMeta: metav1.ObjectMeta{Name: "r-pending"},
+				Status: schedulingv1alpha1.ReservationStatus{
+					Phase: schedulingv1alpha1.ReservationPending,
+				},
+			},
+			expectedMetrics: `
+# HELP scheduler_reservation_status_phase [ALPHA] The current number of reservations in each status phase (e.g. Pending, Available, Succeeded, Failed)
+# TYPE scheduler_reservation_status_phase gauge
+scheduler_reservation_status_phase{name="r-pending",phase="Available"} 0
+scheduler_reservation_status_phase{name="r-pending",phase="Failed"} 0
+scheduler_reservation_status_phase{name="r-pending",phase="Pending"} 1
+scheduler_reservation_status_phase{name="r-pending",phase="Succeeded"} 0
+`,
+		},
+		{
+			name: "Succeeded reservation",
+			reservation: &schedulingv1alpha1.Reservation{
+				ObjectMeta: metav1.ObjectMeta{Name: "r-succeeded"},
+				Status: schedulingv1alpha1.ReservationStatus{
+					Phase: schedulingv1alpha1.ReservationSucceeded,
+				},
+			},
+			expectedMetrics: `
+# HELP scheduler_reservation_status_phase [ALPHA] The current number of reservations in each status phase (e.g. Pending, Available, Succeeded, Failed)
+# TYPE scheduler_reservation_status_phase gauge
+scheduler_reservation_status_phase{name="r-succeeded",phase="Available"} 0
+scheduler_reservation_status_phase{name="r-succeeded",phase="Failed"} 0
+scheduler_reservation_status_phase{name="r-succeeded",phase="Pending"} 0
+scheduler_reservation_status_phase{name="r-succeeded",phase="Succeeded"} 1
+`,
+		},
+		{
+			name: "Failed reservation",
+			reservation: &schedulingv1alpha1.Reservation{
+				ObjectMeta: metav1.ObjectMeta{Name: "r-failed"},
+				Status: schedulingv1alpha1.ReservationStatus{
+					Phase: schedulingv1alpha1.ReservationFailed,
+				},
+			},
+			expectedMetrics: `
+# HELP scheduler_reservation_status_phase [ALPHA] The current number of reservations in each status phase (e.g. Pending, Available, Succeeded, Failed)
+# TYPE scheduler_reservation_status_phase gauge
+scheduler_reservation_status_phase{name="r-failed",phase="Available"} 0
+scheduler_reservation_status_phase{name="r-failed",phase="Failed"} 1
+scheduler_reservation_status_phase{name="r-failed",phase="Pending"} 0
+scheduler_reservation_status_phase{name="r-failed",phase="Succeeded"} 0
+`,
+		},
+		{
+			name: "reservation with empty phase defaults to Pending",
+			reservation: &schedulingv1alpha1.Reservation{
+				ObjectMeta: metav1.ObjectMeta{Name: "r-empty-phase"},
+				Status:     schedulingv1alpha1.ReservationStatus{},
+			},
+			expectedMetrics: `
+# HELP scheduler_reservation_status_phase [ALPHA] The current number of reservations in each status phase (e.g. Pending, Available, Succeeded, Failed)
+# TYPE scheduler_reservation_status_phase gauge
+scheduler_reservation_status_phase{name="r-empty-phase",phase="Available"} 0
+scheduler_reservation_status_phase{name="r-empty-phase",phase="Failed"} 0
+scheduler_reservation_status_phase{name="r-empty-phase",phase="Pending"} 1
+scheduler_reservation_status_phase{name="r-empty-phase",phase="Succeeded"} 0
+`,
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			metrics.Register()
+			metrics.ReservationStatusPhase.Reset()
+
+			RecordReservationPhases(tt.reservation)
+
+			if err := testutil.GatherAndCompare(legacyregistry.DefaultGatherer, strings.NewReader(tt.expectedMetrics), "scheduler_reservation_status_phase"); err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+func TestResyncReservations(t *testing.T) {
+	fakeClientSet := kubefake.NewSimpleClientset()
+	fakeKoordClientSet := koordfake.NewSimpleClientset()
+	sharedInformerFactory := informers.NewSharedInformerFactory(fakeClientSet, 0)
+	koordSharedInformerFactory := koordinformers.NewSharedInformerFactory(fakeKoordClientSet, 0)
+
+	// Register metrics into the global legacyregistry and reset to ensure isolation.
+	metrics.Register()
+	metrics.ReservationStatusPhase.Reset()
+
+	reservations := []*schedulingv1alpha1.Reservation{
+		{
+			ObjectMeta: metav1.ObjectMeta{Name: "r-available"},
+			Status: schedulingv1alpha1.ReservationStatus{
+				Phase: schedulingv1alpha1.ReservationAvailable,
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{Name: "r-pending"},
+			Status: schedulingv1alpha1.ReservationStatus{
+				Phase: schedulingv1alpha1.ReservationPending,
+			},
+		},
+	}
+	for _, r := range reservations {
+		_, err := fakeKoordClientSet.SchedulingV1alpha1().Reservations().Create(context.TODO(), r, metav1.CreateOptions{})
+		assert.NoError(t, err)
+	}
+
+	controller := New(sharedInformerFactory, koordSharedInformerFactory, fakeClientSet, fakeKoordClientSet, &config.ReservationArgs{})
+
+	sharedInformerFactory.Start(nil)
+	koordSharedInformerFactory.Start(nil)
+	sharedInformerFactory.WaitForCacheSync(nil)
+	koordSharedInformerFactory.WaitForCacheSync(nil)
+
+	// First resync: both reservations are written.
+	controller.resyncReservations()
+
+	expectedAfterFirstResync := `
+# HELP scheduler_reservation_status_phase [ALPHA] The current number of reservations in each status phase (e.g. Pending, Available, Succeeded, Failed)
+# TYPE scheduler_reservation_status_phase gauge
+scheduler_reservation_status_phase{name="r-available",phase="Available"} 1
+scheduler_reservation_status_phase{name="r-available",phase="Failed"} 0
+scheduler_reservation_status_phase{name="r-available",phase="Pending"} 0
+scheduler_reservation_status_phase{name="r-available",phase="Succeeded"} 0
+scheduler_reservation_status_phase{name="r-pending",phase="Available"} 0
+scheduler_reservation_status_phase{name="r-pending",phase="Failed"} 0
+scheduler_reservation_status_phase{name="r-pending",phase="Pending"} 1
+scheduler_reservation_status_phase{name="r-pending",phase="Succeeded"} 0
+`
+	if err := testutil.GatherAndCompare(legacyregistry.DefaultGatherer, strings.NewReader(expectedAfterFirstResync), "scheduler_reservation_status_phase"); err != nil {
+		t.Errorf("after first resync: %v", err)
+	}
+
+	// Delete r-available so that the second resync should clean up its metrics via Reset.
+	err := fakeKoordClientSet.SchedulingV1alpha1().Reservations().Delete(context.TODO(), "r-available", metav1.DeleteOptions{})
+	assert.NoError(t, err)
+
+	// Wait until the lister no longer sees r-available (informer cache has processed the delete event).
+	err = wait.PollUntilContextTimeout(context.TODO(), 5*time.Millisecond, 3*time.Second, true, func(ctx context.Context) (bool, error) {
+		list, err := controller.reservationLister.List(labels.Everything())
+		if err != nil {
+			return false, err
+		}
+		for _, r := range list {
+			if r.Name == "r-available" {
+				return false, nil
+			}
+		}
+		return true, nil
+	})
+	assert.NoError(t, err, "timed out waiting for lister to reflect deletion of r-available")
+
+	// Second resync: Reset clears stale r-available metrics, only r-pending remains.
+	controller.resyncReservations()
+
+	expectedAfterSecondResync := `
+# HELP scheduler_reservation_status_phase [ALPHA] The current number of reservations in each status phase (e.g. Pending, Available, Succeeded, Failed)
+# TYPE scheduler_reservation_status_phase gauge
+scheduler_reservation_status_phase{name="r-pending",phase="Available"} 0
+scheduler_reservation_status_phase{name="r-pending",phase="Failed"} 0
+scheduler_reservation_status_phase{name="r-pending",phase="Pending"} 1
+scheduler_reservation_status_phase{name="r-pending",phase="Succeeded"} 0
+`
+	if err := testutil.GatherAndCompare(legacyregistry.DefaultGatherer, strings.NewReader(expectedAfterSecondResync), "scheduler_reservation_status_phase"); err != nil {
+		t.Errorf("after second resync (deleted r-available): %v", err)
+	}
 }
 
 func TestNew(t *testing.T) {

